dipy/tracking/tracker_ptt.pyx

```diff
@@ -0,0 +1,386 @@
+                              * params.inv_voxel_size[i]
+                              + stream_data[7 + i]
+                              * params.ptt.probe_radius
+                              * sin(count * params.ptt.angular_separation)
+                              * params.inv_voxel_size[i])
+
+            stream_data[22] += pmf_gen.get_pmf_value_c(&stream_data[19],
+                                                       &stream_data[1])
+
+
+cdef void prepare_propagator(TrackerParameters params,
+                             double* stream_data,
+                             double arclength) noexcept nogil:
+    """Prepare the propagator.
+
+    The propagator used for transporting the moving frame forward.
+
+    Parameters
+    ----------
+    params : TrackerParameters
+        PTT tracking parameters.
+    stream_data : double*
+        Streamline data persitant across tracking steps.
+    arclength : double
+        Arclenth, which is equivalent to step size along the arc
+
+    """
+    cdef double tmp_arclength
+    stream_data[10] = arclength  # propagator[0]
+
+    if (fabs(stream_data[24]) < params.ptt.k_small
+        and fabs(stream_data[25]) < params.ptt.k_small):
+        stream_data[11] = 0
+        stream_data[12] = 0
+        stream_data[13] = 1
+        stream_data[14] = 0
+        stream_data[15] = 0
+        stream_data[16] = 0
+        stream_data[17] = 0
+        stream_data[18] = 1
+    else:
+        if fabs(stream_data[24]) < params.ptt.k_small:  # k1
+            stream_data[24] = params.ptt.k_small
+        if fabs(stream_data[25]) < params.ptt.k_small:  # k2
+            stream_data[25] = params.ptt.k_small
+
+        tmp_arclength  = arclength * arclength / 2.0
+
+        # stream_data[10:18] -> propagator
+        stream_data[11] = stream_data[24] * tmp_arclength
+        stream_data[12] = stream_data[25] * tmp_arclength
+        stream_data[13] = (1 - stream_data[25]
+                          * stream_data[25] * tmp_arclength
+                          - stream_data[24] * stream_data[24]
+                          * tmp_arclength)
+        stream_data[14] = stream_data[24] * arclength
+        stream_data[15] = stream_data[25] * arclength
+        stream_data[16] = -stream_data[25] * arclength
+        stream_data[17] = (-stream_data[24] * stream_data[25]
+                          * tmp_arclength)
+        stream_data[18] = (1 - stream_data[25] * stream_data[25]
+                          * tmp_arclength)
+
+
+cdef double calculate_data_support(TrackerParameters params,
+                                   double* stream_data,
+                                   PmfGen pmf_gen) noexcept nogil:
+    """Calculates data support for the candidate probe.
+        Parameters
+    ----------
+    params : TrackerParameters
+        PTT tracking parameters.
+    stream_data : double*
+        Streamline data persitant across tracking steps.
+    pmf_gen : PmfGen
+        Orientation data.
```

Comment:
Return value missing.

Reply:
My bad, sorry. There is not return section then. I think there is multiple functions that do not have a return value, so no return section should exist for those.